File: forms/systematic-coronary-risk-evaluation-2-diabetes/full-stack-with-rust-axum-loco-tera-htmx-alpine/src/engine/risk_rules.rs
use super::types::AssessmentData;
use super::utils::{has_established_cvd, hba1c_mmol_mol, calculate_bmi};

pub struct RiskRule {
    pub id: &'static str,
    pub category: &'static str,
    pub description: &'static str,
    pub risk_level: &'static str,
    pub evaluate: fn(&AssessmentData) -> bool,
}

pub fn all_rules() -> Vec<RiskRule> {
    vec![
        // HIGH RISK rules
        RiskRule {
            id: "CVR-001",
            category: "Cardiovascular History",
            description: "Established cardiovascular disease (prior MI, stroke, TIA, PAD, or HF)",
            risk_level: "high",
            evaluate: |d| has_established_cvd(d),
        },
        RiskRule {
            id: "CVR-002",
            category: "Blood Pressure",
            description: "Severely elevated blood pressure (systolic ≥180 mmHg)",
            risk_level: "high",
            evaluate: |d| d.blood_pressure.systolic_bp.is_some_and(|v| v >= 180.0),
        },
        RiskRule {
            id: "CVR-003",
            category: "Glycaemic Control",
            description: "Very poor glycaemic control (HbA1c ≥75 mmol/mol / 9%)",
            risk_level: "high",
            evaluate: |d| hba1c_mmol_mol(d).is_some_and(|v| v >= 75.0),
        },
        RiskRule {
            id: "CVR-004",
            category: "Renal Function",
            description: "Severe renal impairment (eGFR <30 mL/min/1.73m²)",
            risk_level: "high",
            evaluate: |d| d.renal_function.egfr.is_some_and(|v| v < 30.0),
        },
        RiskRule {
            id: "CVR-005",
            category: "Lipid Profile",
            description: "Severely elevated total cholesterol (≥8.0 mmol/L)",
            risk_level: "high",
            evaluate: |d| d.lipid_profile.total_cholesterol.is_some_and(|v| v >= 8.0),
        },
        RiskRule {
            id: "CVR-006",
            category: "Complications",
            description: "Proliferative or maculopathy retinopathy",
            risk_level: "high",
            evaluate: |d| {
                d.complications_screening.retinopathy_status == "proliferative"
                    || d.complications_screening.retinopathy_status == "maculopathy"
            },
        },
        RiskRule {
            id: "CVR-007",
            category: "Diabetes Duration",
            description: "Long diabetes duration (≥20 years)",
            risk_level: "high",
            evaluate: |d| d.diabetes_history.diabetes_duration_years.is_some_and(|v| v >= 20.0),
        },
        // MEDIUM RISK rules
        RiskRule {
            id: "CVR-008",
            category: "Blood Pressure",
            description: "Elevated blood pressure (systolic ≥140 mmHg)",
            risk_level: "medium",
            evaluate: |d| {
                d.blood_pressure.systolic_bp.is_some_and(|v| v >= 140.0 && v < 180.0)
            },
        },
        RiskRule {
            id: "CVR-009",
            category: "Glycaemic Control",
            description: "Above-target HbA1c (≥53 mmol/mol / 7%)",
            risk_level: "medium",
            evaluate: |d| {
                hba1c_mmol_mol(d).is_some_and(|v| v >= 53.0 && v < 75.0)
            },
        },
        RiskRule {
            id: "CVR-010",
            category: "Lipid Profile",
            description: "Elevated LDL cholesterol (≥2.6 mmol/L)",
            risk_level: "medium",
            evaluate: |d| d.lipid_profile.ldl_cholesterol.is_some_and(|v| v >= 2.6),
        },
        RiskRule {
            id: "CVR-011",
            category: "Lifestyle",
            description: "Current smoker",
            risk_level: "medium",
            evaluate: |d| d.lifestyle_factors.smoking_status == "current",
        },
        RiskRule {
            id: "CVR-012",
            category: "Renal Function",
            description: "Moderate renal impairment (eGFR 30-59 mL/min/1.73m²)",
            risk_level: "medium",
            evaluate: |d| d.renal_function.egfr.is_some_and(|v| v >= 30.0 && v < 60.0),
        },
        RiskRule {
            id: "CVR-013",
            category: "Renal Function",
            description: "Albuminuria present (urine ACR ≥3 mg/mmol)",
            risk_level: "medium",
            evaluate: |d| d.renal_function.urine_acr.is_some_and(|v| v >= 3.0),
        },
        RiskRule {
            id: "CVR-014",
            category: "Lifestyle",
            description: "Obesity (BMI ≥30)",
            risk_level: "medium",
            evaluate: |d| {
                let bmi = d.lifestyle_factors.bmi
                    .or_else(|| calculate_bmi(d.patient_demographics.height_cm, d.patient_demographics.weight_kg));
                bmi.is_some_and(|v| v >= 30.0)
            },
        },
        RiskRule {
            id: "CVR-015",
            category: "Cardiovascular",
            description: "Atrial fibrillation",
            risk_level: "medium",
            evaluate: |d| d.cardiovascular_history.atrial_fibrillation == "yes",
        },
        RiskRule {
            id: "CVR-016",
            category: "Cardiovascular",
            description: "Family history of premature cardiovascular disease",
            risk_level: "medium",
            evaluate: |d| d.cardiovascular_history.family_cvd_history == "yes",
        },
        RiskRule {
            id: "CVR-017",
            category: "Lipid Profile",
            description: "Low HDL cholesterol (<1.0 mmol/L male, <1.2 mmol/L female)",
            risk_level: "medium",
            evaluate: |d| {
                let threshold = if d.patient_demographics.sex == "male" { 1.0 } else { 1.2 };
                d.lipid_profile.hdl_cholesterol.is_some_and(|v| v < threshold)
            },
        },
        // LOW RISK rules (positive indicators or minor concerns)
        RiskRule {
            id: "CVR-018",
            category: "Lifestyle",
            description: "Sedentary lifestyle reported",
            risk_level: "low",
            evaluate: |d| d.lifestyle_factors.physical_activity == "sedentary",
        },
        RiskRule {
            id: "CVR-019",
            category: "Complications",
            description: "Background retinopathy present",
            risk_level: "low",
            evaluate: |d| d.complications_screening.retinopathy_status == "background",
        },
        RiskRule {
            id: "CVR-020",
            category: "Lipid Profile",
            description: "Elevated triglycerides (≥2.3 mmol/L)",
            risk_level: "low",
            evaluate: |d| d.lipid_profile.triglycerides.is_some_and(|v| v >= 2.3),
        },
    ]
}

File: forms/systematic-coronary-risk-evaluation-2-diabetes/front-end-form-with-svelte/src/lib/engine/risk-rules.ts
import type { AssessmentData, FiredRule } from './types.js';
import { hasEstablishedCvd, hba1cMmolMol, calculateBmi } from './utils.js';

export interface RiskRule {
	id: string;
	category: string;
	description: string;
	riskLevel: string;
	evaluate: (d: AssessmentData) => boolean;
}

export function allRules(): RiskRule[] {
	return [
		// HIGH RISK rules
		{
			id: 'CVR-001',
			category: 'Cardiovascular History',
			description: 'Established cardiovascular disease (prior MI, stroke, TIA, PAD, or HF)',
			riskLevel: 'high',
			evaluate: (d) => hasEstablishedCvd(d)
		},
		{
			id: 'CVR-002',
			category: 'Blood Pressure',
			description: 'Severely elevated blood pressure (systolic >= 180 mmHg)',
			riskLevel: 'high',
			evaluate: (d) => d.bloodPressure.systolicBp != null && d.bloodPressure.systolicBp >= 180
		},
		{
			id: 'CVR-003',
			category: 'Glycaemic Control',
			description: 'Very poor glycaemic control (HbA1c >= 75 mmol/mol / 9%)',
			riskLevel: 'high',
			evaluate: (d) => {
				const v = hba1cMmolMol(d);
				return v != null && v >= 75;
			}
		},
		{
			id: 'CVR-004',
			category: 'Renal Function',
			description: 'Severe renal impairment (eGFR < 30 mL/min/1.73m2)',
			riskLevel: 'high',
			evaluate: (d) => d.renalFunction.egfr != null && d.renalFunction.egfr < 30
		},
		{
			id: 'CVR-005',
			category: 'Lipid Profile',
			description: 'Severely elevated total cholesterol (>= 8.0 mmol/L)',
			riskLevel: 'high',
			evaluate: (d) =>
				d.lipidProfile.totalCholesterol != null && d.lipidProfile.totalCholesterol >= 8.0
		},
		{
			id: 'CVR-006',
			category: 'Complications',
			description: 'Proliferative or maculopathy retinopathy',
			riskLevel: 'high',
			evaluate: (d) =>
				d.complicationsScreening.retinopathyStatus === 'proliferative' ||
				d.complicationsScreening.retinopathyStatus === 'maculopathy'
		},
		{
			id: 'CVR-007',
			category: 'Diabetes Duration',
			description: 'Long diabetes duration (>= 20 years)',
			riskLevel: 'high',
			evaluate: (d) =>
				d.diabetesHistory.diabetesDurationYears != null &&
				d.diabetesHistory.diabetesDurationYears >= 20
		},
		// MEDIUM RISK rules
		{
			id: 'CVR-008',
			category: 'Blood Pressure',
			description: 'Elevated blood pressure (systolic >= 140 mmHg)',
			riskLevel: 'medium',
			evaluate: (d) =>
				d.bloodPressure.systolicBp != null &&
				d.bloodPressure.systolicBp >= 140 &&
				d.bloodPressure.systolicBp < 180
		},
		{
			id: 'CVR-009',
			category: 'Glycaemic Control',
			description: 'Above-target HbA1c (>= 53 mmol/mol / 7%)',
			riskLevel: 'medium',
			evaluate: (d) => {
				const v = hba1cMmolMol(d);
				return v != null && v >= 53 && v < 75;
			}
		},
		{
			id: 'CVR-010',
			category: 'Lipid Profile',
			description: 'Elevated LDL cholesterol (>= 2.6 mmol/L)',
			riskLevel: 'medium',
			evaluate: (d) =>
				d.lipidProfile.ldlCholesterol != null && d.lipidProfile.ldlCholesterol >= 2.6
		},
		{
			id: 'CVR-011',
			category: 'Lifestyle',
			description: 'Current smoker',
			riskLevel: 'medium',
			evaluate: (d) => d.lifestyleFactors.smokingStatus === 'current'
		},
		{
			id: 'CVR-012',
			category: 'Renal Function',
			description: 'Moderate renal impairment (eGFR 30-59 mL/min/1.73m2)',
			riskLevel: 'medium',
			evaluate: (d) =>
				d.renalFunction.egfr != null &&
				d.renalFunction.egfr >= 30 &&
				d.renalFunction.egfr < 60
		},
		{
			id: 'CVR-013',
			category: 'Renal Function',
			description: 'Albuminuria present (urine ACR >= 3 mg/mmol)',
			riskLevel: 'medium',
			evaluate: (d) => d.renalFunction.urineAcr != null && d.renalFunction.urineAcr >= 3
		},
		{
			id: 'CVR-014',
			category: 'Lifestyle',
			description: 'Obesity (BMI >= 30)',
			riskLevel: 'medium',
			evaluate: (d) => {
				const bmi =
					d.lifestyleFactors.bmi ??
					calculateBmi(d.patientDemographics.heightCm, d.patientDemographics.weightKg);
				return bmi != null && bmi >= 30;
			}
		},
		{
			id: 'CVR-015',
			category: 'Cardiovascular',
			description: 'Atrial fibrillation',
			riskLevel: 'medium',
			evaluate: (d) => d.cardiovascularHistory.atrialFibrillation === 'yes'
		},
		{
			id: 'CVR-016',
			category: 'Cardiovascular',
			description: 'Family history of premature cardiovascular disease',
			riskLevel: 'medium',
			evaluate: (d) => d.cardiovascularHistory.familyCvdHistory === 'yes'
		},
		{
			id: 'CVR-017',
			category: 'Lipid Profile',
			description: 'Low HDL cholesterol (< 1.0 mmol/L male, < 1.2 mmol/L female)',
			riskLevel: 'medium',
			evaluate: (d) => {
				const threshold = d.patientDemographics.sex === 'male' ? 1.0 : 1.2;
				return (
					d.lipidProfile.hdlCholesterol != null &&
					d.lipidProfile.hdlCholesterol < threshold
				);
			}
		},
		// LOW RISK rules
		{
			id: 'CVR-018',
			category: 'Lifestyle',
			description: 'Sedentary lifestyle reported',
			riskLevel: 'low',
			evaluate: (d) => d.lifestyleFactors.physicalActivity === 'sedentary'
		},
		{
			id: 'CVR-019',
			category: 'Complications',
			description: 'Background retinopathy present',
			riskLevel: 'low',
			evaluate: (d) => d.complicationsScreening.retinopathyStatus === 'background'
		},
		{
			id: 'CVR-020',
			category: 'Lipid Profile',
			description: 'Elevated triglycerides (>= 2.3 mmol/L)',
			riskLevel: 'low',
			evaluate: (d) =>
				d.lipidProfile.triglycerides != null && d.lipidProfile.triglycerides >= 2.3
		}
	];
}

/** Evaluate all rules against assessment data. */
export function evaluateRules(data: AssessmentData): FiredRule[] {
	const fired: FiredRule[] = [];
	for (const rule of allRules()) {
		if (rule.evaluate(data)) {
			fired.push({
				id: rule.id,
				category: rule.category,
				description: rule.description,
				riskLevel: rule.riskLevel
			});
		}
	}
	return fired;
}
